finat/finiteelementbase.py

```diff
@@ -144,6 +145,112 @@ def point_evaluation(self, order, refcoords, entity=None):
+        and a alpha tensor for extracting the alpha components from ReferenceGrad,
+        sorted by total derivative order.
+
+        For example, a dual basis containing 2 functionals with maximum derivative
+        order of 1 would be represented by:
+
+        .. highlight:: python
+        .. code-block:: python
+
+            ((((point_set_10, weight_tensor_10, alpha_tensor_10),
+               (point_set_11, weight_tensor_11, alpha_tensor_11)), None)
+             (((point_set_20, weight_tensor_20, alpha_tensor_20),
+              ()), None)
+
+        where one of the innermost tuples is empty because there are no evaluations
+        at that order. The `None` is either a placeholder for \'normal\' elements,
+        or a tuple containing the index of expression to extract using gem.Delta
+        (currently just used by TensorFiniteElement).
+        '''
+        # TODO: Add label for type of evaluation?
+
+    def dual_evaluation(self, fn, entity=None):
+        '''Return code for performing the dual evaluation at the nodes of the
+        reference element. Currently only works for non-derivatives.
+
+        :param fn: Callable that takes in PointSet and returns GEM expression.
+        :param entity: the cell entity on which to tabulate for comparing
+                       results with FIAT.
+        '''
+        if entity is not None:
+            # TODO: Add comparison to FIAT
+            raise NotImplementedError("Comparison with FIAT is not implemented!")
+
+        if any(len(dual) > 1 for dual, tensorfe_idx in self.dual_basis):
+            raise NotImplementedError("Can only interpolate onto dual basis functionals"
+                                      " without derivative evaluation, sorry!")
+
+        dual_expressions = []   # One for each functional
+        expr_cache = {}         # Sharing of evaluation of the expression at points
+
+        for dual, tensorfe_idx in self.dual_basis:
+            qexprs = gem.Zero()
+            for derivative_order, deriv in enumerate(dual):
+                for tups in deriv:
+                    try:
+                        point_set, weight_tensor, alpha_tensor = tups
+                    except ValueError:  # Empty
+                        continue
+
+                    try:
+                        # TODO: Choose hash method
+                        expr = expr_cache[(point_set.points.data.tobytes(), alpha_tensor.array.tobytes())]
+                    except KeyError:
+                        expr_grad = fn(point_set, derivative=derivative_order)
+                        # TODO: Multiple alpha at once
+                        if derivative_order == 0:
+                            expr = expr_grad
+                        else:
+                            # Extract derivative component
+                            alpha_idx = tuple(gem.Index(extent=fn.dimension) for _ in range(derivative_order))
```

Comment:
This requires that `fn` satisfies a particular interface which the docstring does not advertise.

Reply:
Added requirement for dimension attribute to docstring.